fix 좀비프로세스 생기는 오류 수정

zhy2on opened the pull request:
- kqueue에 fork한 자식프로세스 pid의 exit 이벤트를 모니터링할 수 있게 등록
- 이벤트가 발생하면 waitpid WNOHANG 옵션을 이용해 논블럭킹으로 자식 프로세스의 종료 상태를 회수함

zhy2on committed:
fix: 좀비프로세스 생기는 오류 수정
- kqueue에 fork한 자식프로세스 pid의 exit 이벤트를 모니터링할 수 있게 등록
- 이벤트가 발생하면 waitpid WNOHANG 옵션을 이용해 논블럭킹으로 자식 프로세스의 종료 상태를 회수함

aLVINlEE9 reviewed (approved):
LGTM

zhy2on committed:
fix: env_list_ 할당 오류 수정
- cgi_envs_.size() + 1 만큼 할당해줘야 함
- env_list_ 마지막을 NULL로 채워줘야 하기 때문

minjune8506 commented on sources/core/webserv.cpp:
```diff
@@ -61,6 +61,10 @@ void Webserv::RunServer() {
 		struct kevent event = kq_handler_.MonitorEvent(); // get event
 		ClientSocket *client = FindClientSocket(event.ident);
+		if (event.fflags & NOTE_EXIT) {
+			waitpid(event.ident, 0, WNOHANG);
```
여기서 waitpid 안하고 바로 끝내버리면 똑같은 문제가 발생하나요?

minjune8506 replied:
kqueue 설명 보면은 data에 wait 결과값이 저장되어 나온다고 되어있는데
그럼 kqueue 내부적으로 프로세스 종료를 기다리는거..?

zhy2on replied:
저도 그래서 원랜 이벤트 등록만 해두고 waitpid도 안 했는데
그랬더니 똑같이 좀비 프로세스가 생기더라구요..
그래서 wait이나 waitpid를 써야 부모프로세스가 회수를 했다는 걸 아는 건가? 해서 추가해봤더니 사라지더라고요🤔

zhy2on replied:
프로세스가 종료된 것도 알 수 있고 data에 종료상태가 저장되어 나오는 것도 맞는데
wait관련 함수를 써야 부모 프로세스가 자식 프로세스의 종료상태를 회수했다고 생각하는 게 아닌가.. 싶습니다

minjune8506 replied:
이게 맞는거같아요 wait pid를 써야될거같네요 WNOHANG 옵션 있어서 어짜피 blocking이 안되는거니까..

minjune8506 replied:
waitpid 함수 한번 보시고 result랑 두번째 인자 이용해서 예외처리만 하면 될 것 같습니다.

zhy2on replied:
엇 예외처리해서 저건데 어떻게 수정하는 거 말씀하시는 걸까요 ??

zhy2on replied:
아 반환상태가 오류에 관한 상태일 경우 예외처리 말씀하시는 걸까요?

minjune8506 replied:
네네 그것도 있고 waitpid 두번째 인자로 변수하나 넣으면 자식 프로세스가 정상 / 비정상 종료 여부를 알 수 있다네요

zhy2on replied:
아하 넵넵 처리해서 다시 올리도록 하겠습니다

zhy2on replied:
확인 한 번만 부탁드립니다!

zhy2on committed:
feat: waitpid 관련 예외처리 추가
    - waitpid 자체가 오류일 시 예외처리
    - 자식프로세스가 정상종료가 아닌 경우 예외처
    - cgi의 반환값이 0이 아닌 경우(execve가 실패했던 경우 or cgi자체의 리턴값이 오류였던 경우) 예외처리

minjune8506 reviewed (approved):
LGTM
